source/agora/common/Metadata.d

```diff
@@ -0,0 +1,132 @@
+
+    Defines metadata which is stored to disk (e.g. peer info)
+
+    Copyright:
+        Copyright (c) 2019 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.common.Metadata;
+
+import agora.common.Data;
+import agora.common.Set;
+
+import vibe.core.log;
+import vibe.data.json;
+
+import std.array;
+import std.file;
+import std.format;
+import std.meta;
+import std.stdio;
+import std.typecons;
+
+/// Metadata stored to disk
+public class Metadata
+{
+    /// Known validators we want to try re-connecting to
+    public Set!Address validators;
+
+    /// Known listeners we may want to try re-connecting to
+    public Set!Address listeners;
+
+    /// Path to the data dir
+    private const string file_name;
+
+    /// Whether metadata was loaded on startup (and validators are not empty)
+    private bool loaded;
+
+    /// During unittesting we shouldn't do any file I/O
+    version(unittest) Appender!string test_appender;
+
+
+    /// Constructor
+    public this ( string data_dir )
+    {
+        this.file_name = format("%s/%s", data_dir, "metadata.dat");
+        version(unittest) this.test_appender = appender!string();
+    }
+
+    /// Dump the metadata to disk
+    public void dump ( )
+    {
+        version(unittest)
+        {
+            alias output = test_appender;
+        }
+        else
+        {
+            scope file = File(this.file_name, "w");
+            auto output = file.lockingTextWriter();
+        }
+
+        serializeToJson(output, tuple(validators, listeners));
+    }
+
+    /// Load any existing metadata from disk (if it exists),
+    /// and set 'this.loaded' to true if at least one validator was loaded
+    public void load ( )
+    {
+        version (unittest)
+        {
+            auto raw_data = this.test_appender.data;
+            if (raw_data.length == 0)
+                return;
+        }
+        else
+        {
+            if (!this.file_name.exists)
+                return;
+
+            scope file = File(this.file_name, "r");
+            string raw_data = cast(string)std.file.read(file_name);
+        }
+
+        auto data = deserializeJson!(Tuple!(Set!Address, Set!Address))(raw_data);
+
+        this.validators = data[0];
+        this.listeners = data[1];
+
+        // there must be at least one validator present in the metadata,
+        // otherwise we have to discover network from the 'network' confing section
+        this.loaded = this.validators.length > 0;
+    }
+
+    /// Returns: true if the metadata was successfully loaded from disk
+    public bool hasLoaded ( )
+    {
+        return this.loaded;
+    }
+}
+
+///
+unittest
+{
+    auto meta = new Metadata(tempDir());
+    const Validator = "http://127.0.0.1:1234";
+    const Listener = "http://127.0.0.1:4567";
```

Comment:
Yeah that's not scalable. Since we serialize to `json`, you should be able to do:
```d
unittest
{
    auto meta = new Metadata("{ ... }");
}
```

The API here does not matter to much though. You might want to make your ctor non-validating and provide a `static Metadata fromFile(string path)` function as well as a `static Metadata fromString(string)`.

Reply:
Alrighty, I'll put these as helper functions outside the Metadata class.